cloudsql/mysql/database-sql/cloudsql.go

```diff
@@ -0,0 +1,299 @@
+	// connection pool will be created instead.
+	if os.Getenv("DB_TCP_HOST") != "" {
+		db, err = initTcpConnectionPool()
+		if err != nil {
+			log.Fatalf("initTcpConnectionPool: unable to connect: %s", err)
+		}
+	} else {
+		db, err = initSocketConnectionPool()
+		if err != nil {
+			log.Fatalf("initSocketConnectionPool: unable to connect: %s", err)
+		}
+	}
+
+	// Create the votes table if it does not already exist.
+	if _, err = db.Exec(`CREATE TABLE IF NOT EXISTS votes
+	( vote_id SERIAL NOT NULL, time_cast timestamp NOT NULL,
+	candidate CHAR(6) NOT NULL, PRIMARY KEY (vote_id) );`); err != nil {
+		log.Fatalf("unable to create table: %s", err)
+	}
+
+	http.HandleFunc("/", indexHandler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+// indexHandler handles requests to the / route.
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		err := showTotals(w, r)
+		if err != nil {
+			log.Fatalf("indexHandler: error returned from showTotals: %s", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	case "POST":
+		saveVote(w, r)
+	default:
+		http.Error(w, fmt.Sprintf("HTTP Method %s Not Allowed", r.Method), http.StatusMethodNotAllowed)
+	}
+}
+
+// recentVotes returns a slice of the last 5 votes cast.
+func recentVotes() ([]vote, error) {
+	var votes []vote
+	rows, err := db.Query(`SELECT candidate, time_cast FROM votes ORDER BY time_cast DESC LIMIT 5`)
+	if err != nil {
+		return votes, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		nextVote := vote{}
+		rowError := rows.Scan(&nextVote.Candidate, &nextVote.VoteTime)
+		if err != nil {
+			return votes, rowError
+		}
+		votes = append(votes, nextVote)
+	}
+	if err = rows.Err(); err != nil {
+		return votes, err
+	}
+	return votes, nil
+}
+
+// currentTotals returns a templateData structure for populating the web page.
+func currentTotals() (templateData, error) {
+
+	// get total votes for each candidate
+	var tabVotes, spaceVotes uint
+	err := db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='TABS'`).Scan(&tabVotes)
+	if err != nil {
+		return templateData{}, err
+	}
+	err = db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='SPACES'`).Scan(&spaceVotes)
+	if err != nil {
+		return templateData{}, err
+	}
+
+	var voteDiffStr string = voteDiff(int(math.Abs(float64(tabVotes) - float64(spaceVotes)))).String()
+
+	latestVotesCast, err := recentVotes()
+	if err != nil {
+		return templateData{}, err
+	}
+	return templateData{tabVotes, spaceVotes, voteDiffStr, latestVotesCast}, nil
+
+}
+
+// showTotals renders an HTML template showing the current vote totals.
+func showTotals(w http.ResponseWriter, r *http.Request) error {
+
+	totals, err := currentTotals()
+	if err != nil {
+		return err
```

Comment:
When relaying errors, prepend them with the returning function name so the error message carries the stack:

```suggestion
		return fmt.Errorf("currentTotals: %v", err)
```

This should be carried through all errors.